pkg/terraform/exec/plugins/plugin_versions_generate.go

```diff
@@ -0,0 +1,249 @@
+// +build tools
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/module"
+)
+
+type TFPluginInfo struct {
+	Module  module.Version
+	Name    string
+	Tag     string
+	Version string
+}
+
+type TFPlugins struct {
+	Pkg          string
+	TFPluginInfo []TFPluginInfo
+}
+
+func getTemplate() string {
+	return `// Code generated by plugin_versions_generate.go; DO NOT EDIT.
+
+package {{ .Pkg }}
+
+{{ with .TFPluginInfo }}
+        {{- range . }}
+
+func Get{{ .Name }}Version() string {
+	return "{{ .Version }}"
+}
+
+        {{- end }}
+{{- end}}
+`
+}
+
+// TODO: local plugins (azureprivatedns, vsphereprivate)
+
+// getTerraformPluginInfo finds terraform providers used in go.mod, does any
+// necessary replacements and saves into our plugin map.
+func getTerraformPluginInfo(file *modfile.File, plugins *map[string]*TFPluginInfo) {
+	tre := regexp.MustCompile(`^[[:alnum:]][[:alnum:]\.\-/]+terraform-provider`)
+	nre := regexp.MustCompile(`(terraform-provider-([[:alnum:]]+))`)
+
+	// Find all terraform providers
+	for _, r := range file.Require {
+		if tre.Match([]byte(r.Mod.Path)) {
+			matches := nre.FindStringSubmatch(r.Mod.Path)
+			if len(matches) < 3 {
+				log.Fatal("failed to parse ", r.Mod.Path)
+			}
+			(*plugins)[matches[1]] = &TFPluginInfo{
+				Module: r.Mod,
+				Name:   matches[2],
+			}
+		}
+	}
+
+	// Update replaced terraform providers
+	for _, r := range file.Replace {
+		if tre.Match([]byte(r.Old.Path)) {
+			matches := nre.FindStringSubmatch(r.Old.Path)
+			if len(matches) < 3 {
+				log.Fatal("failed to parse ", r.Old.Path)
+			}
+			(*plugins)[matches[1]] = &TFPluginInfo{
+				Module: r.New,
+				Name:   matches[2],
+			}
+		}
+	}
+}
+
+// getTerraformPluginVersions gathers the terraform provider version numbers
+func getTerraformPluginVersions(file *modfile.File, plugins *map[string]*TFPluginInfo) {
+	hre := regexp.MustCompile(`([a-fA-F0-9]{12})$`)
+	ore := regexp.MustCompile(`[[:space:]]+|\^[[:digit:]]+`)
+	pre := regexp.MustCompile(`terraform-provider-[[:alnum:]]+/v[[:digit:]]+`)
+	tre := regexp.MustCompile(`[[:digit:]]{14}-([a-fA-F0-9]{12})$`)
+	ure := regexp.MustCompile(`^(.+terraform-provider-[[:alnum:]]+)`)
+	vre := regexp.MustCompile(`([0-9\.]+)`)
+
+	for p, pi := range *plugins {
+		var repo *git.Repository
+		var mod *module.Version = &pi.Module
+
+		// XXX: we assume https checkouts
+		url := fmt.Sprintf("https://%s", mod.Path)
+
+		// If this is a v2 or up  module, chop off the end
+		if pre.Match([]byte(mod.Path)) {
+			path := ure.FindString(mod.Path)
+			url = fmt.Sprintf("https://%s", path)
+		}
```

Comment:
I think we can combine the regular expressions together so we only have to do one match rather than two separate matches, and I think it makes the regex more expressive.

`^(.+terraform-provider-[[:alnum:]]+)(/v[[:digit:]]+)?`

```suggestion
		submatches := newRegex.FindStringSubmatch(mod.Path)
		url = fmt.Sprintf("https://%s", submatches[1])
```

Reply:
I have simplified all of the regular expressions and minimized them, I think we are good here now.